<html>

<body bgcolor="#C0C0C0">

<h2>The Gene Designer Menu </h2>
<table class="image">
<tr><td><img src="menu_image.png" alt="Gene Designer Menu"/></td></tr>
<tr><td><caption align="bottom">Gene Designer Menu on OS X</caption></td></tr>
</table>
<p><a href="help.html">Back to Main Help Navigation Page</a><p>
<h3>Menu Options </h3>
<ul>
<li><a href="#window">Window</a></li>
<li><a href="#file">File</a></li>
<li><a href="#reference">Reference</a></li>
<li><a href="#databases">Databases</a></li>
<li><a href="#tools">Tools</a></li>
<li><a href="#guide">Guide</a></li>
<li><a href="#help">Help</a></li>
</ul>


<h4><a name="window">Window</a></h4>
<p>
Default options for controlling positioning of window, including minimization.  Shortcut for minimization on OS X is Cmd-M.
</p>


<h4><a name="file">File</a></h4>
<dl>
<dt>Undo</dt>
<dd>Restores value of currently selected text box to a previous state.  Works for Sequence, Vector, and Amino Acid Sequence text boxes.  Shortcut: Cmd-Z.</dd>
<dt>Redo</dt>
<dd>Undo an undo. Shortcut: Cmd-Y.</dd>
<dt>About</dt>
<dd>Find out more about Gene Designer.</dd>
<dt>Quit</dt>
<dd>Exit the program. Shortcut: Cmd-Q.</dd>
</dl>


<h4><a name="reference">Reference</a></h4>
<dl><dt>Enzyme Reference</dt>
<dd>Opens a window displaying a list of restriction enzymes and the target pattern at which they cut.  Consult the REBASE database for where the enzyme cuts in relation to the sequence.  Click the thumbnail for an image:
<a href="enz_ref.html"><img src="table_thumb.png" alt="Enzyme Reference Table" align="bottom" width="20" height="10"/></a></dd>
<dt>Codon Reference</dt><dd>Opens a window displaying a list of codons and their associated amino acid translations.  Click the thumbnail for an image:
<a href="codon_ref.html"><img src="table_thumb.png" alt="Codon Reference Table" align="bottom" width="20" height="10"/></a></dd>
<dt>Commercial Reference</dt><dd>Opens a window with a table enzymes and vendors according to the REBASE database.  A legend associating each vendor with a letter is located at the top of the window.  Click the thumbnail for an image:
<a href="comm_ref.html"><img src="table_thumb.png" alt="Commercial Reference Table and Legend"j align="bottom" width="20" height="10"/></a></dd>


<h4><a name="databases">Databases</a></h4>
<dl>
<dt>Restriction Enzymes--up to date as of 10/6/10</dt>
	<dd>Add Restriction Enzyme--creates a panel for you to enter the enzyme name, the pattern the enzyme recognizes, methylation sites, and commercial suppliers.  Information should be entered as done in the REBASE databases: http://rebase.neb.com/rebase</dd>
	<dd>Delete Restriction Enzyme--deletes the restriction enzyme currently selected in the enzyme list.  This is a permanent change!  If an enzyme is accidentally deleted, you can re-enter the information through the "Add Restriction Enzyme" option.</dd>
<dt>Vectors</dt>
	<dd>Add Vector--creates a panel for the addition of a vector to the database.  Enter the vector name and vector sequence in the boxes provided.  Alternatively, a vector may be added from a file in FASTA format.</dd>
	<dd>Delete Vector--deletes the vector currently selected in the vector list</dd>
<dt>Codon Usage</dt>
	<dd>Add Organism--creates a panel where you may enter an organism name and a nucleotide sequence.  Gene Designer will compute the frequency of each codon in the sequence and store it in the database.  You may also enter codon frequency percentages (out of 1000) if known.</dd>
	<dd>Delete Organism--deletes the currently selected organism in the menu below.  This is a permanent change!  E. coli K-12 is not allowed to be deleted.</dd>
	<dd>Submenu--list of organisms whose codon frequencies can be used by Gene Designer in optimizing sequences or in making mutations to sequences.  E. coli K-12 is the default but other organisms can be added through the add organism option.</dd>
<dt>Templates</dt>
	<dd>Template Sequence Database--displays a list of all sequences already added to the template database.  Clicking the load sequence button loads the currently highlighted sequence into the Mutation Panel and Optimization Panel displays.  Click the thumbnail for an image:
	<a href="load_template.html"><img src="table_thumb.png" alt="Load Template Display" align="bottom" width="20" height="10"/></a></dd>
	<dd>Add New Template Sequences--User selects a FASTA file containing sequences.  The user is allowed to rename sequences before submission to the database.  Sequences are stored permanently to make it easy to looad for future use.
	<a href="new_template.html"><img src="table_thumb.png" alt="New Template Sequence Display" align="bottom" width="20" height="10"/></a></dd>
</dl>

<h4><a name="tools">Tools</a></h4>
<dl><dt>Check Sequence/Host Toxicity</dt>
<dd>Reports whether any codons in the sequence entered into the sequence box represent less than 1% of the total codons of the "host" organism.  The host organism is the selected organism on the "Codon Usage" submenu found under the databases menu.  By
default E. coli K-12 is the selected host. Shortcut: Cmd-T</dd>
<dt>Find Open Reading Frame</dt>
<dd>Finds open reading frames (ORFs) of a minimum length (determined by the number of amino acids between the start and stop codons) specified by the user in a pop-up dialog.  The start and end positions of all ORFs	meeting the criteria are displayed along with the ORF nucleotide sequence.  A pop-up will inform the user if no 
ORFs satisfy the requirements.</dd>
<dt>Homology Modeling</dt>
<dd>User selects a FASTA file containing protein sequences with template sequence listed first.  The PDB file for the template sequence should be in the same directory as the FASTA file.  An alignment of the sequences is created and Modeller is used to create a homology model.  The resulting model is displayed using PyMol.</dd>
</dl>

<h4><a name="guide">Guide</a></h4>
<p>
Launches the Help Panel, but it seems you have already figured that out.
</p>


<h4><a name="help">Help</a></h4>
<p>
Search for help on OS X.  I am not sure if this is also on Windows.
</p>


</body>